src/QirRuntime/experimental/IonqTranslator.cpp

```diff
@@ -0,0 +1,664 @@
+        if (v.size() > 1)
+        {
+            os << "[";
+        }
+        for (long l : v)
+        {
+            os << l << ",";
+        }
+        if (v.size() > 1)
+        {
+            os << "],";
+        }
+    }
+
+    /*==============================================================================
+        As the simulator executes along a particular path, the translator would
+        accumulate corresponding instructions and errors.
+    ==============================================================================*/
+    struct TranslationError
+    {
+        Gate gate = Gate::undefined;
+        vector<long> targets;  // can be empty if error isn't specific to a particular qubit
+        vector<long> controls; // can be empty
+        string msg;
+
+        void Print(ostringstream& os) const
+        {
+            os << "{";
+            os << "\"gate\":\"" << gate_names[static_cast<size_t>(gate)] << "\",";
+            PrintVector(os, targets, "target");
+            PrintVector(os, controls, "control");
+            os << "\"message\":\"" << msg << "\"";
+            os << "},";
+        }
+    };
+
+    struct Instruction
+    {
+        Gate gate;
+        vector<long> targets;
+        vector<long> controls; // can be empty
+        double rotation = 0;   // only meaningful for rx, ry, rz, xx, yy and zz gates
+
+        Instruction(Gate gate, vector<long>&& targets, vector<long>&& controls, double rotation)
+            : gate(gate)
+            , targets(std::move(targets))
+            , controls(std::move(controls))
+            , rotation(rotation)
+        {
+        }
+
+        bool IsRotation() const
+        {
+            return Gate::rx <= gate && gate <= Gate::zz;
+        }
+
+        void Print(ostringstream& os) const
+        {
+            os << "{";
+            os << "\"gate\":\"" << gate_names[static_cast<size_t>(gate)] << "\",";
+            PrintVector(os, targets, "target");
+            PrintVector(os, controls, "control");
+            if (IsRotation())
+            {
+                os << "\"rotation\":" << rotation << ",";
+            }
+            os << "},";
+        }
+    };
+
+    /*==============================================================================
+        The decomposer provides classical "text book" decompositions for some of the
+        unitaries without any hardware specific optimizations.
+
+        A quantum processor, that wants to avail of the decompositions, should
+        implement the primitive operations, pass itself to the decomposer and then
+        delegate to it calls that should be decomposed.
+
+        NB: because the owner opts-in for each decomposition separately the decomposer
+        will call back for each step, even if it might come back to it (R1Frac -> R1).
+    ==============================================================================*/
+    class CCanonicalDecomposer final : public CQuantumApiBase

```

Comment:
I'd like to keep this example for now as it demonstrates how IQuantumApi can be used to layer/reuse partial implementations. Given it's in experimental folder and we are not planning to provide a "real" IonQ target here any time soon I'm not worried it will be in the way of the new patterns for decomposition.

Per your other comment, I'll rename the simulator to avoid IonQ reference and make it clear it's just an exercise in using the layering + ITranslator.